### DISCUSSION TITLE
role based login for two diffrent panels

### DISCUSSION THREAD

**Discussion by mohandtharsha** [filamentphp/filament#8477] [Help]
Hi I have two panel one is admin panel and employer and if user role is admin it should login to admin dashboard else login to employer dashboard and admin cannot login to employer dashboard and employer cannot login into admin dashboard for this I have created custom middleware to protect role based login ,

Below is checkUserMiddleware.php,

[<?php

namespace App\Http\Middleware;

use Closure;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Auth;
use Symfony\Component\HttpFoundation\Response;
use App\Models\Roles;
use Illuminate\Validation\ValidationException;
use App\Exceptions\PermissionDeniedException;
use Illuminate\Support\Facades\Redirect;
use Illuminate\Support\Facades\Session;

class CheckUserRole
{
    public function handle(Request $request, Closure $next, $role)
    {
        if (Auth::check()) {
            $roleName = ltrim($role, ':');
            $authUser = Auth::user();

            if ($authUser) {
                $user = $authUser->getAttributes();
                $role_id = $user['role_id'];
                $role = Roles::where('id', $role_id)->first();
                $role_name = $role->name;

                if ($roleName === $role_name) {
                    // Allow access for the specific user
                    return $next($request);
                } else {
                    // return response()->redirectTo('employ', 302); // Use 302 for temporary redirectionSession::flash("error", "You do not have permission to access this resource.");
                    // return redirect()->route('employer.login');

                    // Session::flash("success", "Changes applied successfully");
                    // return redirect('employer/login');
                    // return Redirect::back()->withErrors(['error' => 'You do not have permission to access this resource.']);
                    // return back();
                    // Auth::logout();
                    // return response()->json(['error' => 'You do not have permission to access this resource.'], 403);
                   // Catch the custom exception
                //    throw new PermissionDeniedException('You do not have permission to login');
                //    return Redirect::back()->withErrors(['error' => 'You do not have permission to access this resource.']);

                // return Redirect::to('/employer/login');

                }
            }
        }

        return $next($request);
    }

}

Below is employer panel provider,

<?php

namespace App\Providers\Filament;

use Filament\Http\Middleware\Authenticate;
use Filament\Http\Middleware\DisableBladeIconComponents;
use Filament\Http\Middleware\DispatchServingFilamentEvent;
use Filament\Pages;
use Filament\Panel;
use Filament\PanelProvider;
use Filament\Support\Colors\Color;
use Filament\Widgets;
use Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse;
use Illuminate\Cookie\Middleware\EncryptCookies;
use Illuminate\Foundation\Http\Middleware\VerifyCsrfToken;
use Illuminate\Routing\Middleware\SubstituteBindings;
use Illuminate\Session\Middleware\AuthenticateSession;
use Illuminate\Session\Middleware\StartSession;
use Illuminate\View\Middleware\ShareErrorsFromSession;
use Filament\Navigation\MenuItem;
use App\Filament\Employer\Pages\EmployersProfile;
use App\Filament\Employer\Resources\EmployerUsersResource;
use pxlrbt\FilamentSpotlight\SpotlightPlugin;

class EmployerPanelProvider extends PanelProvider
{
    public function panel(Panel $panel): Panel
    {
        return $panel
            ->id('employer')
            ->path('employer')
            ->login()

            // ->profile()
            ->colors([
                'primary' => Color::Blue,
            ])
            ->discoverResources(in: app_path('Filament/Employer/Resources'), for: 'App\\Filament\\Employer\\Resources')
            ->discoverPages(in: app_path('Filament/Employer/Pages'), for: 'App\\Filament\\Employer\\Pages')
            ->pages([
                Pages\Dashboard::class,
            ])
            ->discoverWidgets(in: app_path('Filament/Employer/Widgets'), for: 'App\\Filament\\Employer\\Widgets')
            ->widgets([
                Widgets\AccountWidget::class,
                Widgets\FilamentInfoWidget::class,
            ])
            ->middleware([
                EncryptCookies::class,
                AddQueuedCookiesToResponse::class,
                StartSession::class,
                AuthenticateSession::class,
                ShareErrorsFromSession::class,
                VerifyCsrfToken::class,
                SubstituteBindings::class,
                DisableBladeIconComponents::class,
                'checkUserRole::employer',
            ])

            ->authMiddleware([
                Authenticate::class,
            ])
            ->sidebarCollapsibleOnDesktop()
            ->userMenuItems([
                MenuItem::make()
                    ->label('Profile')
                    ->url(fn (): string => EmployersProfile::getUrl())
                    ->icon('heroicon-o-user-circle'),
                // MenuItem::make()
                //     ->label('Users')
                //     ->url(fn (): string => EmployerUsersResource::getUrl())
                //     ->icon('heroicon-o-users'),
                MenuItem::make()
                    ->label('Messages')
                    // ->url(fn (): string => EmployerUsersResource::getUrl())
                    ->icon('heroicon-o-folder'),
                MenuItem::make()
                    ->label('Settings')
                    // ->url(fn (): string => EmployerUsersResource::getUrl())
                    ->icon('heroicon-o-cog-8-tooth'),
                // ...
            ]);
    }
}

Below is my kernel.php 

  <?php
  
  namespace App\Http;
  
  use Illuminate\Foundation\Http\Kernel as HttpKernel;
  
  class Kernel extends HttpKernel
  {
      /**
       * The application's global HTTP middleware stack.
       *
       * These middleware are run during every request to your application.
       *
       * @var array<int, class-string|string>
       */
      protected $middleware = [
          // \App\Http\Middleware\TrustHosts::class,
          \App\Http\Middleware\TrustProxies::class,
          \Illuminate\Http\Middleware\HandleCors::class,
          \App\Http\Middleware\PreventRequestsDuringMaintenance::class,
          \Illuminate\Foundation\Http\Middleware\ValidatePostSize::class,
          \App\Http\Middleware\TrimStrings::class,
          \Illuminate\Foundation\Http\Middleware\ConvertEmptyStringsToNull::class,
      ];
  
      /**
       * The application's route middleware groups.
       *
       * @var array<string, array<int, class-string|string>>
       */
      protected $middlewareGroups = [
          'web' => [
              \App\Http\Middleware\EncryptCookies::class,
              \Illuminate\Cookie\Middleware\AddQueuedCookiesToResponse::class,
              \Illuminate\Session\Middleware\StartSession::class,
              \Illuminate\View\Middleware\ShareErrorsFromSession::class,
              \App\Http\Middleware\VerifyCsrfToken::class,
              \Illuminate\Routing\Middleware\SubstituteBindings::class,
          ],
  
          'api' => [
              // \Laravel\Sanctum\Http\Middleware\EnsureFrontendRequestsAreStateful::class,
              \Illuminate\Routing\Middleware\ThrottleRequests::class.':api',
              \Illuminate\Routing\Middleware\SubstituteBindings::class,
          ],
      ];
  
      /**
       * The application's middleware aliases.
       *
       * Aliases may be used instead of class names to conveniently assign middleware to routes and groups.
       *
       * @var array<string, class-string|string>
       */
      protected $middlewareAliases = [
          'auth' => \App\Http\Middleware\Authenticate::class,
          'auth.basic' => \Illuminate\Auth\Middleware\AuthenticateWithBasicAuth::class,
          'auth.session' => \Illuminate\Session\Middleware\AuthenticateSession::class,
          'cache.headers' => \Illuminate\Http\Middleware\SetCacheHeaders::class,
          'can' => \Illuminate\Auth\Middleware\Authorize::class,
          'guest' => \App\Http\Middleware\RedirectIfAuthenticated::class,
          'password.confirm' => \Illuminate\Auth\Middleware\RequirePassword::class,
          'precognitive' => \Illuminate\Foundation\Http\Middleware\HandlePrecognitiveRequests::class,
          'signed' => \App\Http\Middleware\ValidateSignature::class,
          'throttle' => \Illuminate\Routing\Middleware\ThrottleRequests::class,
          'verified' => \Illuminate\Auth\Middleware\EnsureEmailIsVerified::class,
      ];
  
      protected $routeMiddleware = [
          'checkUserRole' => \App\Http\Middleware\CheckUserRole::class,
      ];
  }

**Comment by nsommer89** [comment]
You can add the following in User model:
```
public function canAccessPanel(Panel $panel): bool
    {
        if (!$this->hasVerifiedEmail()) {
            return false;
        }

        switch ($panel->getId()) {
            case 'admin':
                return $this->hasRole('Super Admin');
            case 'organizer':
                return $this->hasRole('Organizer');
            default:
                return false;
        }
    }
````

And as well, add the following use statement .... 
`
use Filament\Panel;
`
